File: backend/app/schemas/produccion.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.fechas import ModeloConFechasUtc


class ConsumoRolloProduccion(BaseModel):
    rollo_id: int
    metros: float = Field(gt=0)


class StockAdicionalItemCrear(BaseModel):
    """Una línea de unidades adicionales que quedaron disponibles más allá
    de lo que el apartado necesitaba (o, en producción libre, simplemente
    unidades que se quieren dejar en inventario). `calidad` clasifica cada
    línea por separado para poder tener, por ejemplo, 2 de Primera y 1 de
    Segunda con su propio motivo en la misma producción."""
    calidad: Literal["primera", "segunda"]
    cantidad: int = Field(gt=0)
    motivo_segunda: str = ""

    @model_validator(mode="after")
    def _exigir_motivo_si_segunda(self) -> "StockAdicionalItemCrear":
        if self.calidad == "segunda" and not self.motivo_segunda.strip():
            raise ValueError("Indica el motivo/descripción del defecto para una unidad de Segunda.")
        return self


class ProduccionCrear(BaseModel):
    # "teja" (default, todo el comportamiento existente) o un producto
    # "seccionado" (caballete/flanche): se corta distinto — cada corte de N
    # metros de largo del rollo produce SIEMPRE la misma cantidad de
    # unidades (el ancho del rollo se divide siempre en ese número de
    # secciones — 3 para caballete, 5 para flanche), así que
    # `cantidad_productos` + `metros_por_unidad` (la longitud) determinan
    # cuántos cortes hacen falta y cuánto sobra — ver
    # app/services/produccion.py::SECCIONES_POR_TIPO_PRODUCTO y
    # registrar_produccion.
    tipo_producto: Literal["teja", "caballete", "flanche"] = "teja"
    producto_fabricado: str = ""
    modelo: str
    medida_producto: str
    cantidad_productos: int = Field(gt=0)
    responsable: str
    observaciones: str = ""
    rollos: list[ConsumoRolloProduccion]
    # Si viene de una solicitud de producción (apartado enviado a producción);
    # cuando se informa, la cotización se toma del apartado y no se puede editar.
    apartado_item_id: int | None = None

    # Datos del lote, para las unidades de `stock_adicional` (si las hay) —
    # ninguno es obligatorio si no se registra stock adicional. Para
    # caballetes, `metros_por_unidad` (la longitud) es SIEMPRE obligatorio:
    # sin ella no se puede calcular cuántos cortes hacen falta.
    color: str = ""
    ral: str = ""
    calibre: str = ""
    metros_por_unidad: float | None = None
    stock_adicional: list[StockAdicionalItemCrear] = []

    @model_validator(mode="after")
    def _exigir_metros_por_unidad_si_hace_falta(self) -> "ProduccionCrear":
        if self.tipo_producto == "teja" and not self.calibre.strip():
            raise ValueError("Indica el calibre — es obligatorio para identificar y consolidar correctamente el stock de TEJA.")
        if self.tipo_producto in ("caballete", "flanche") and not self.metros_por_unidad:
            raise ValueError(f"Indica la longitud del {self.tipo_producto} (metros por unidad) para calcular los cortes.")
        if self.stock_adicional and not self.metros_por_unidad:
            raise ValueError("Indica los metros por unidad para registrar el stock adicional.")
        return self


class RolloUtilizadoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    rollo_id: int
    identificador_rollo: str
    metros_consumidos: float


class StockAdicionalGeneradoResponse(BaseModel):
    """Una unidad de stock (Producto) que quedó disponible por esta
    producción — para que la confirmación en pantalla muestre exactamente
    qué código/cantidad/calidad se generó."""
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: str
    descripcion: str
    stock: float
    calidad: str | None = None
    motivo_segunda: str = ""
    color: str = ""
    ral: str = ""
    calibre: str = ""
    metros_por_unidad: float | None = None
    codigo_rollo_origen: str = ""
    ancho_rollo: float | None = None
    tipo_producto: str = ""


class ProduccionResponse(ModeloConFechasUtc):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo_unico: str
    cotizacion: str
    fecha: datetime
    usuario: str
    responsable: str
    bodega_id: int
    tipo_producto: str = "teja"
    producto_fabricado: str
    modelo: str
    medida_producto: str
    cantidad_productos: int
    codigo_clasificacion: str
    total_metros_consumidos: float
    saldo_codigo: float
    metros_excedente: float = 0
    observaciones: str
    apartado_item_id: int | None = None
    cliente_apartado: str = ""
    rollos_utilizados: list[RolloUtilizadoResponse] = []
    productos_stock: list[StockAdicionalGeneradoResponse] = []

File: src/Utils/produccion.ts
/** Regla física de los productos "seccionados": el ancho del rollo se
 * divide siempre en N partes iguales según el tipo, así que cada corte
 * produce SIEMPRE esas N unidades. Un solo lugar (reutilizado por el
 * controlador de Producción y por Inventario) define cuántas secciones
 * tiene cada tipo — agregar un producto nuevo de esta misma familia física
 * es sumar una entrada aquí, no repetir el número por toda la app. Debe
 * coincidir exactamente con SECCIONES_POR_TIPO_PRODUCTO del backend
 * (app/services/produccion.py). */
export const SECCIONES_POR_TIPO_PRODUCTO: Record<string, number> = { caballete: 3, flanche: 5 };
export const NOMBRE_POR_TIPO_PRODUCTO: Record<string, string> = { caballete: "caballete", flanche: "flanche" };

/** `ancho_rollo` (m) de un stock, dividido por el número de secciones que
 * le corresponde según su `tipoProducto` — o null si no aplica (tejas y
 * cualquier producto que no venga de un tipo seccionado). */
export function anchoPorSeccion(anchoRollo: number | null | undefined, tipoProducto: string) {
  const secciones = SECCIONES_POR_TIPO_PRODUCTO[tipoProducto];
  if (anchoRollo == null || !secciones) return null;
  return Math.round((anchoRollo / secciones) * 10000) / 10000;
}

type ProductoStockItem = {
  id: number | string; calidad?: string; motivoSegunda?: string;
  metrosPorUnidad?: number | null; stock: number; codigoRolloOrigen?: string;
};
type RolloUtilizado = { identificadorRollo: string };
type ProduccionParaStock = {
  id: number | string; codigoUnico: string; fecha: string; responsable: string;
  productoFabricado: string; modelo: string; medidaProducto: string;
  codigoClasificacion: string; saldoCodigo: number;
  productosStock?: ProductoStockItem[]; rollosUtilizados?: RolloUtilizado[];
};

/** Cada unidad de stock adicional que generó una producción se muestra como
 * una fila aparte en el historial (Hoja de Vida / Producciones registradas),
 * nunca fusionada con la fila de la cotización real: mismo código de
 * producción que la fila de arriba, pero con Cotización = "STOCK" y
 * Cliente = "-", para que quede claro que esa unidad no salió con el
 * cliente de esa cotización, sino que quedó disponible en inventario. */
export function filasStockAdicional(prod: ProduccionParaStock) {
  if (!prod.productosStock || prod.productosStock.length === 0) return [];
  const referencia = (prod.rollosUtilizados || []).map((r) => r.identificadorRollo).join(", ");
  return prod.productosStock.map((item) => {
    // Genérico a propósito: no asume "teja" ni ningún otro producto — la
    // columna Producto/Modelo de la misma fila ya dice qué se fabricó.
    let observaciones = item.calidad === "segunda" ? "Segunda" : "Primera";
    if (item.calidad === "segunda" && item.motivoSegunda) observaciones += `: ${item.motivoSegunda}`;
    const metrosConsumidos = item.metrosPorUnidad != null
      ? Math.round(item.stock * item.metrosPorUnidad * 100) / 100
      : null;
    return {
      key: `${prod.id}-stock-${item.id}`,
      codigoProduccion: prod.codigoUnico,
      cotizacion: "STOCK",
      cliente: "-",
      fecha: prod.fecha,
      responsable: prod.responsable,
      productoFabricado: prod.productoFabricado,
      modelo: prod.modelo,
      medida: prod.medidaProducto,
      cantidad: item.stock,
      codigoClasificacion: prod.codigoClasificacion,
      referencia,
      metrosConsumidos,
      saldoRestante: prod.saldoCodigo,
      codigoRolloOrigen: item.codigoRolloOrigen,
      observaciones,
    };
  });
}
